grpc-client-utils/src/main/java/org/hypertrace/core/grpcutils/client/GrpcChannelRegistry.java

```diff
@@ -50,7 +66,55 @@ private String getChannelId(String host, int port, boolean isPlaintext) {
   }
 
   public void shutdown() {
-    channelMap.values().forEach(ManagedChannel::shutdown);
+    this.shutdown(this.clock.instant().plus(1, MINUTES));
+  }
+
+  public void shutdown(Instant deadline) {
```

Comment:
Can we add an explanatory comment about the entire shutdown strategy? Would be really good for a new reader, as to why did we follow this route

Reply:
I omitted the part about leaking and resources because I'm still not sure it's actually right, more a theory - did (hopefully) thoroughly explained how the shutdown works though.